src/features/surveys/models/SurveyDataModel.ts

```diff
@@ -20,7 +30,107 @@ export default class SurveyDataModel extends ModelBase {
     return this._repo.getSurvey(this._orgId, this._surveyId);
   }
 
+  publish(): void {
+    const { data } = this.getData();
+    if (!data) {
+      return;
+    }
+
+    const now = dayjs();
+    const today = now.format('YYYY-MM-DD');
+
+    const { published, expires } = data;
+
+    if (!published && !expires) {
+      this._repo.updateSurvey(this._orgId, this._surveyId, {
+        published: today,
+      });
+    } else if (!published) {
+      const endDate = dayjs(expires);
+      if (endDate.isBefore(today)) {
+        this._repo.updateSurvey(this._orgId, this._surveyId, {
+          expires: null,
+          published: today,
+        });
+      } else if (endDate.isAfter(today)) {
+        this._repo.updateSurvey(this._orgId, this._surveyId, {
+          published: today,
+        });
+      }
+    } else if (!expires) {
+      // Start date is non-null
+      const startDate = dayjs(published);
+      if (startDate.isAfter(today)) {
+        // End date is null, start date is future
+        this._repo.updateSurvey(this._orgId, this._surveyId, {
+          published: today,
+        });
+      }
+    } else {
+      // Start and end date are non-null
+      const startDate = dayjs(published);
+      const endDate = dayjs(expires);
+
+      if (
+        (startDate.isBefore(today) || startDate.isSame(today)) &&
+        (endDate.isBefore(today) || endDate.isSame(today))
+      ) {
+        // Start is past, end is past
+        this._repo.updateSurvey(this._orgId, this._surveyId, {
+          expires: null,
+        });
+      } else if (startDate.isAfter(today) && endDate.isAfter(today)) {
+        // Start is future, end is future
+        this._repo.updateSurvey(this._orgId, this._surveyId, {
+          published: today,
+        });
+      }
+    }
+  }
+
   setTitle(title: string) {
     this._repo.updateSurvey(this._orgId, this._surveyId, { title });
   }
+
+  get state(): SurveyState {
+    const { data } = this.getData();
+    if (!data) {
+      return SurveyState.UNKNOWN;
+    }
+
+    if (data.published) {
+      const publishDate = new Date(data.published);
+      const now = new Date();
+
+      if (publishDate > now) {
+        return SurveyState.SCHEDULED;
+      } else {
+        if (data.expires) {
+          const expiryDate = new Date(data.expires);
+
+          if (expiryDate < now) {
+            return SurveyState.UNPUBLISHED;
```

Comment:
Cool! I almost wrote exactly same as urs except for unpublished part. So I have no objection haha XD